### PR TITLE
[ᚬhuobi] [feat] add riscv built-in service

### PR DESCRIPTION
**Pull request by zeroqn**
**What type of PR is this?**
feat

**What this PR does / why we need it**:
add riscv built-in service

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -90,3 +89,29 @@ security-audit:
 .PHONY: build prod prod-test
 .PHONY: fmt test clippy doc doc-deps doc-api check stats
 .PHONY: ci info security-audit
+.PHONY: test_riscv
+.PHONY: duktape
+
+test_riscv:
```
Comment (huwenchao) [inline Makefile]: This is for test usage, can be deleted.

File: built-in-services/riscv/examples/dex/riscv_demo.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    res = await query(erc20_addr, {method: 'balance_of'});
+    assert(!res.isError);
+    assert(res.ret == '999999701');
+    res = await send_tx(dex_addr, {method: 'withdraw', asset: erc20_addr, amount: 200});
+    assert(res.includes('[ProtocolError]'));
+    res = await query(dex_addr, {method: 'balance_of', asset: erc20_addr});
+    assert(!res.isError);
+    assert(res.ret == '199');
+}
+main();
```
Comment (huwenchao) [inline built-in-services/riscv/examples/dex/riscv_demo.ts]: new line

File: built-in-services/riscv/src/vm/c/pvm.h [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+// Params:
+//   addr: a pointer to a buffer in VM memory space denoting where the address located at.
+// Return:
+//   code: 0(success)
+int pvm_caller(uint8_t *addr)
+{
+    return syscall(SYSCODE_CALLER, addr, 0, 0, 0, 0, 0);
+}
+
+int pvm_address(uint8_t *addr)
```
Comment (huwenchao) [inline built-in-services/riscv/src/vm/c/pvm.h]: miss docs

File: Makefile (hunk 1 above)
Comment (zeroqn) [reply Makefile]: Done

File: built-in-services/riscv/examples/dex/riscv_demo.ts (hunk 2 above)
Comment (zeroqn) [reply built-in-services/riscv/examples/dex/riscv_demo.ts]: Done

File: built-in-services/riscv/src/vm/c/pvm.h (hunk 3 above)
Comment (zeroqn) [reply built-in-services/riscv/src/vm/c/pvm.h]: Done

File: built-in-services/riscv/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    fn deploy(
+        &mut self,
+        ctx: ServiceContext,
+        payload: DeployPayload,
+    ) -> ProtocolResult<DeployResp> {
+        let code = Bytes::from(hex::decode(&payload.code).map_err(ServiceError::HexDecode)?);
+        let code_hash = Hash::digest(code.clone());
+        // FIXME:
+        // let code_len = code.len() as u64;
+        // ctx.sub_cycles(code_len)?;
```
[the remark this answers is not in the document]
Comment (zeroqn) [reply built-in-services/riscv/src/lib.rs]: All tests pass. Done

**Pull request merged**